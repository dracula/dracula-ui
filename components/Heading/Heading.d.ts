import React, { HTMLAttributes } from 'react';
import { MarginMixin, PaddingMixin } from '../../base/spacing';
export declare const headingSizes: {
    'heading-1': string;
    'heading-2': string;
    'heading-3': string;
    'heading-4': string;
    'heading-5': string;
    'heading-6': string;
};
export declare const headingColors: {
    purpleCyan: string;
    yellowPink: string;
    cyanGreen: string;
    pinkPurple: string;
    animated: string;
    cyan: string;
    green: string;
    orange: string;
    pink: string;
    purple: string;
    red: string;
    yellow: string;
    white: string;
    black: string;
    blackSecondary: string;
    blackLight: string;
};
/** Heading Props */
export interface HeadingProps extends HTMLAttributes<HTMLHeadingElement>, MarginMixin, PaddingMixin {
    /**
     * The size and type of Heading to be used.
     * Denotes hierarchy.
     */
    size?: keyof typeof headingSizes;
    /**
     * The Dracula UI color to be applied to the Heading.
     */
    color?: keyof typeof headingColors;
    as?: keyof HTMLElementTagNameMap;
}
/**
 * Heading is used to display headlines and other forms of hierarchical Text.
 *
 * Headings are similar to the base Text component, but restricted to certain
 * sizes and font weight.
 */
export declare const Heading: React.FC<HeadingProps>;
