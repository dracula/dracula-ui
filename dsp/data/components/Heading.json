{
  "class": "component",
  "type": "page",
  "id": "Heading",
  "name": "Heading",
  "last_updated": "2021-03-14T22:34:30.073Z",
  "last_updated_by": "System",
  "description": "Heading is used to display headlines and other forms of hierarchical Text.\n\nHeadings are similar to the base Text component, but restricted to certain\nsizes and font weight.",
  "related_entity_ids": [],
  "tags": ["component"],
  "snippets": {
    "trigger": "heading",
    "languages": {
      "javascriptreact": "<Heading>Heading</Heading>;\n",
      "html": "<h1\n  class=\"drac-heading drac-heading-1 drac-text-white drac-spacing-none drac-margin-none\"\n>\n  Heading\n</h1>\n"
    }
  },
  "ext_com_draculaui_variations": [
    {
      "title": "Usage",
      "html": "<h1\n  class=\"drac-heading drac-heading-1 drac-text-white drac-spacing-none drac-margin-none\"\n>\n  Heading\n</h1>\n",
      "react": "<Heading>Heading</Heading>;\n",
      "docs": "",
      "screenshot": "./dsp/assets/Heading.png",
      "svgData": "./dsp/assets/svgs/Heading.svg"
    },
    {
      "title": "Sizes",
      "html": "<div>\n  <div style=\"margin-bottom: 20px\">\n    <h1\n      color=\"white\"\n      class=\"drac-heading drac-heading-1 drac-text-white drac-spacing-none drac-margin-none\"\n    >\n      heading-1\n    </h1>\n  </div>\n  <div style=\"margin-bottom: 20px\">\n    <h2\n      color=\"white\"\n      class=\"drac-heading drac-heading-2 drac-text-white drac-spacing-none drac-margin-none\"\n    >\n      heading-2\n    </h2>\n  </div>\n  <div style=\"margin-bottom: 20px\">\n    <h3\n      color=\"white\"\n      class=\"drac-heading drac-heading-3 drac-text-white drac-spacing-none drac-margin-none\"\n    >\n      heading-3\n    </h3>\n  </div>\n  <div style=\"margin-bottom: 20px\">\n    <h4\n      color=\"white\"\n      class=\"drac-heading drac-heading-4 drac-text-white drac-spacing-none drac-margin-none\"\n    >\n      heading-4\n    </h4>\n  </div>\n  <div style=\"margin-bottom: 20px\">\n    <h5\n      color=\"white\"\n      class=\"drac-heading drac-heading-5 drac-text-white drac-spacing-none drac-margin-none\"\n    >\n      heading-5\n    </h5>\n  </div>\n  <div style=\"margin-bottom: 20px\">\n    <h6\n      color=\"white\"\n      class=\"drac-heading drac-heading-6 drac-text-white drac-spacing-none drac-margin-none\"\n    >\n      heading-6\n    </h6>\n  </div>\n</div>\n",
      "react": "<div>\n  <div style={{ marginBottom: 20 }}>\n    <Heading color=\"white\" size=\"heading-1\">\n      heading-1\n    </Heading>\n  </div>\n  <div style={{ marginBottom: 20 }}>\n    <Heading color=\"white\" size=\"heading-2\">\n      heading-2\n    </Heading>\n  </div>\n  <div style={{ marginBottom: 20 }}>\n    <Heading color=\"white\" size=\"heading-3\">\n      heading-3\n    </Heading>\n  </div>\n  <div style={{ marginBottom: 20 }}>\n    <Heading color=\"white\" size=\"heading-4\">\n      heading-4\n    </Heading>\n  </div>\n  <div style={{ marginBottom: 20 }}>\n    <Heading color=\"white\" size=\"heading-5\">\n      heading-5\n    </Heading>\n  </div>\n  <div style={{ marginBottom: 20 }}>\n    <Heading color=\"white\" size=\"heading-6\">\n      heading-6\n    </Heading>\n  </div>\n</div>;\n",
      "docs": "",
      "screenshot": "./dsp/assets/HeadingSizes.png",
      "svgData": "./dsp/assets/svgs/HeadingSizes.svg"
    },
    {
      "title": "Colors",
      "html": "<div>\n  <div style=\"margin-bottom: 20px\">\n    <h1\n      color=\"white\"\n      class=\"drac-heading drac-heading-1 drac-text-white drac-spacing-none drac-margin-none\"\n    >\n      white\n    </h1>\n  </div>\n  <div style=\"margin-bottom: 20px\">\n    <h1\n      color=\"black\"\n      class=\"drac-heading drac-heading-1 drac-text-black drac-spacing-none drac-margin-none\"\n    >\n      black\n    </h1>\n  </div>\n  <div style=\"margin-bottom: 20px\">\n    <h1\n      color=\"blackSecondary\"\n      class=\"drac-heading drac-heading-1 drac-text-black-secondary drac-spacing-none drac-margin-none\"\n    >\n      blackSecondary\n    </h1>\n  </div>\n  <div style=\"margin-bottom: 20px\">\n    <h1\n      color=\"blackLight\"\n      class=\"drac-heading drac-heading-1 drac-text-black-light drac-spacing-none drac-margin-none\"\n    >\n      blackLight\n    </h1>\n  </div>\n  <div style=\"margin-bottom: 20px\">\n    <h1\n      color=\"cyan\"\n      class=\"drac-heading drac-heading-1 drac-text-cyan drac-spacing-none drac-margin-none\"\n    >\n      cyan\n    </h1>\n  </div>\n  <div style=\"margin-bottom: 20px\">\n    <h1\n      color=\"green\"\n      class=\"drac-heading drac-heading-1 drac-text-green drac-spacing-none drac-margin-none\"\n    >\n      green\n    </h1>\n  </div>\n  <div style=\"margin-bottom: 20px\">\n    <h1\n      color=\"orange\"\n      class=\"drac-heading drac-heading-1 drac-text-orange drac-spacing-none drac-margin-none\"\n    >\n      orange\n    </h1>\n  </div>\n  <div style=\"margin-bottom: 20px\">\n    <h1\n      color=\"pink\"\n      class=\"drac-heading drac-heading-1 drac-text-pink drac-spacing-none drac-margin-none\"\n    >\n      pink\n    </h1>\n  </div>\n  <div style=\"margin-bottom: 20px\">\n    <h1\n      color=\"purple\"\n      class=\"drac-heading drac-heading-1 drac-text-purple drac-spacing-none drac-margin-none\"\n    >\n      purple\n    </h1>\n  </div>\n  <div style=\"margin-bottom: 20px\">\n    <h1\n      color=\"red\"\n      class=\"drac-heading drac-heading-1 drac-text-red drac-spacing-none drac-margin-none\"\n    >\n      red\n    </h1>\n  </div>\n  <div style=\"margin-bottom: 20px\">\n    <h1\n      color=\"yellow\"\n      class=\"drac-heading drac-heading-1 drac-text-yellow drac-spacing-none drac-margin-none\"\n    >\n      yellow\n    </h1>\n  </div>\n  <div style=\"margin-bottom: 20px\">\n    <h1\n      color=\"purpleCyan\"\n      class=\"drac-heading drac-heading-1 drac-text-purple-cyan drac-spacing-none drac-margin-none\"\n    >\n      purpleCyan\n    </h1>\n  </div>\n  <div style=\"margin-bottom: 20px\">\n    <h1\n      color=\"yellowPink\"\n      class=\"drac-heading drac-heading-1 drac-text-yellow-pink drac-spacing-none drac-margin-none\"\n    >\n      yellowPink\n    </h1>\n  </div>\n  <div style=\"margin-bottom: 20px\">\n    <h1\n      color=\"cyanGreen\"\n      class=\"drac-heading drac-heading-1 drac-text-cyan-green drac-spacing-none drac-margin-none\"\n    >\n      cyanGreen\n    </h1>\n  </div>\n  <div style=\"margin-bottom: 20px\">\n    <h1\n      color=\"pinkPurple\"\n      class=\"drac-heading drac-heading-1 drac-text-pink-purple drac-spacing-none drac-margin-none\"\n    >\n      pinkPurple\n    </h1>\n  </div>\n  <div style=\"margin-bottom: 20px\">\n    <h1\n      color=\"animated\"\n      class=\"drac-heading drac-heading-1 drac-text-animated drac-spacing-none drac-margin-none\"\n    >\n      animated\n    </h1>\n  </div>\n</div>\n",
      "react": "<div>\n  <div style={{ marginBottom: 20 }}>\n    <Heading color=\"white\">white</Heading>\n  </div>\n  <div style={{ marginBottom: 20 }}>\n    <Heading color=\"black\">black</Heading>\n  </div>\n  <div style={{ marginBottom: 20 }}>\n    <Heading color=\"blackSecondary\">blackSecondary</Heading>\n  </div>\n  <div style={{ marginBottom: 20 }}>\n    <Heading color=\"blackLight\">blackLight</Heading>\n  </div>\n  <div style={{ marginBottom: 20 }}>\n    <Heading color=\"cyan\">cyan</Heading>\n  </div>\n  <div style={{ marginBottom: 20 }}>\n    <Heading color=\"green\">green</Heading>\n  </div>\n  <div style={{ marginBottom: 20 }}>\n    <Heading color=\"orange\">orange</Heading>\n  </div>\n  <div style={{ marginBottom: 20 }}>\n    <Heading color=\"pink\">pink</Heading>\n  </div>\n  <div style={{ marginBottom: 20 }}>\n    <Heading color=\"purple\">purple</Heading>\n  </div>\n  <div style={{ marginBottom: 20 }}>\n    <Heading color=\"red\">red</Heading>\n  </div>\n  <div style={{ marginBottom: 20 }}>\n    <Heading color=\"yellow\">yellow</Heading>\n  </div>\n  <div style={{ marginBottom: 20 }}>\n    <Heading color=\"purpleCyan\">purpleCyan</Heading>\n  </div>\n  <div style={{ marginBottom: 20 }}>\n    <Heading color=\"yellowPink\">yellowPink</Heading>\n  </div>\n  <div style={{ marginBottom: 20 }}>\n    <Heading color=\"cyanGreen\">cyanGreen</Heading>\n  </div>\n  <div style={{ marginBottom: 20 }}>\n    <Heading color=\"pinkPurple\">pinkPurple</Heading>\n  </div>\n  <div style={{ marginBottom: 20 }}>\n    <Heading color=\"animated\">animated</Heading>\n  </div>\n</div>;\n",
      "docs": "",
      "screenshot": "./dsp/assets/HeadingColors.png",
      "svgData": "./dsp/assets/svgs/HeadingColors.svg"
    }
  ],
  "ext_com_draculaui_props": {
    "size": {
      "defaultValue": null,
      "description": "The size and type of Heading to be used.\nDenotes hierarchy.",
      "name": "size",
      "parent": {
        "fileName": "src/components/Heading/Heading.tsx",
        "name": "HeadingProps"
      },
      "declarations": [
        {
          "fileName": "src/components/Heading/Heading.tsx",
          "name": "HeadingProps"
        }
      ],
      "required": false,
      "type": {
        "name": "\"heading-1\" | \"heading-2\" | \"heading-3\" | \"heading-4\" | \"heading-5\" | \"heading-6\" | undefined"
      }
    },
    "color": {
      "defaultValue": null,
      "description": "The Dracula UI color to be applied to the Heading.",
      "name": "color",
      "parent": {
        "fileName": "src/components/Heading/Heading.tsx",
        "name": "HeadingProps"
      },
      "declarations": [
        {
          "fileName": "src/components/Heading/Heading.tsx",
          "name": "HeadingProps"
        }
      ],
      "required": false,
      "type": {
        "name": "\"white\" | \"black\" | \"blackSecondary\" | \"blackLight\" | \"cyan\" | \"green\" | \"orange\" | \"pink\" | \"purple\" | \"red\" | \"yellow\" | \"purpleCyan\" | \"yellowPink\" | \"cyanGreen\" | \"pinkPurple\" | \"animated\" | undefined"
      }
    },
    "spacing": {
      "defaultValue": null,
      "description": "Controls the spacing between the Heading component and its parent and siblings.",
      "name": "spacing",
      "parent": {
        "fileName": "src/components/Heading/Heading.tsx",
        "name": "HeadingProps"
      },
      "declarations": [
        {
          "fileName": "src/components/Heading/Heading.tsx",
          "name": "HeadingProps"
        }
      ],
      "required": false,
      "type": { "name": "SpacingPropType | undefined" }
    },
    "margin": {
      "defaultValue": null,
      "description": "Dracula UI standard margin properties.",
      "name": "margin",
      "parent": {
        "fileName": "src/components/Heading/Heading.tsx",
        "name": "HeadingProps"
      },
      "declarations": [
        {
          "fileName": "src/components/Heading/Heading.tsx",
          "name": "HeadingProps"
        }
      ],
      "required": false,
      "type": { "name": "MarginPropType | undefined" }
    },
    "as": {
      "defaultValue": null,
      "description": "",
      "name": "as",
      "parent": {
        "fileName": "src/components/Heading/Heading.tsx",
        "name": "HeadingProps"
      },
      "declarations": [
        {
          "fileName": "src/components/Heading/Heading.tsx",
          "name": "HeadingProps"
        }
      ],
      "required": false,
      "type": { "name": "keyof HTMLElementTagNameMap | undefined" }
    }
  },
  "ext_com_draculaui_docgen": {
    "tags": {},
    "description": "Heading is used to display headlines and other forms of hierarchical Text.\n\nHeadings are similar to the base Text component, but restricted to certain\nsizes and font weight.",
    "displayName": "Heading",
    "methods": [],
    "props": {
      "size": {
        "defaultValue": null,
        "description": "The size and type of Heading to be used.\nDenotes hierarchy.",
        "name": "size",
        "parent": {
          "fileName": "src/components/Heading/Heading.tsx",
          "name": "HeadingProps"
        },
        "declarations": [
          {
            "fileName": "src/components/Heading/Heading.tsx",
            "name": "HeadingProps"
          }
        ],
        "required": false,
        "type": {
          "name": "\"heading-1\" | \"heading-2\" | \"heading-3\" | \"heading-4\" | \"heading-5\" | \"heading-6\" | undefined"
        }
      },
      "color": {
        "defaultValue": null,
        "description": "The Dracula UI color to be applied to the Heading.",
        "name": "color",
        "parent": {
          "fileName": "src/components/Heading/Heading.tsx",
          "name": "HeadingProps"
        },
        "declarations": [
          {
            "fileName": "src/components/Heading/Heading.tsx",
            "name": "HeadingProps"
          }
        ],
        "required": false,
        "type": {
          "name": "\"white\" | \"black\" | \"blackSecondary\" | \"blackLight\" | \"cyan\" | \"green\" | \"orange\" | \"pink\" | \"purple\" | \"red\" | \"yellow\" | \"purpleCyan\" | \"yellowPink\" | \"cyanGreen\" | \"pinkPurple\" | \"animated\" | undefined"
        }
      },
      "spacing": {
        "defaultValue": null,
        "description": "Controls the spacing between the Heading component and its parent and siblings.",
        "name": "spacing",
        "parent": {
          "fileName": "src/components/Heading/Heading.tsx",
          "name": "HeadingProps"
        },
        "declarations": [
          {
            "fileName": "src/components/Heading/Heading.tsx",
            "name": "HeadingProps"
          }
        ],
        "required": false,
        "type": { "name": "SpacingPropType | undefined" }
      },
      "margin": {
        "defaultValue": null,
        "description": "Dracula UI standard margin properties.",
        "name": "margin",
        "parent": {
          "fileName": "src/components/Heading/Heading.tsx",
          "name": "HeadingProps"
        },
        "declarations": [
          {
            "fileName": "src/components/Heading/Heading.tsx",
            "name": "HeadingProps"
          }
        ],
        "required": false,
        "type": { "name": "MarginPropType | undefined" }
      },
      "as": {
        "defaultValue": null,
        "description": "",
        "name": "as",
        "parent": {
          "fileName": "src/components/Heading/Heading.tsx",
          "name": "HeadingProps"
        },
        "declarations": [
          {
            "fileName": "src/components/Heading/Heading.tsx",
            "name": "HeadingProps"
          }
        ],
        "required": false,
        "type": { "name": "keyof HTMLElementTagNameMap | undefined" }
      }
    }
  }
}
