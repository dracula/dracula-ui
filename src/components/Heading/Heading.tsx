import cx from 'classnames/dedupe'
import mapValues from 'lodash/mapValues'
import React, { HTMLAttributes } from 'react'
import { colors } from '../../base/colors'
import {
  cleanProps,
  marginMixin,
  MarginMixin,
  paddingMixin,
  PaddingMixin
} from '../../base/spacing'

export const headingSizes = {
  '2xl': 'drac-heading-2xl',
  xl: 'drac-heading-xl',
  lg: 'drac-heading-lg',
  md: 'drac-heading',
  sm: 'drac-heading-sm',
  xs: 'drac-heading-xs'
}

export const headingColors = mapValues(colors, (className) => {
  return className.replace('-bg-', '-text-')
})

/** Heading Props */
export interface HeadingProps
  extends HTMLAttributes<HTMLHeadingElement>,
    MarginMixin,
    PaddingMixin {
  /**
   * The size and type of Heading to be used.
   * Denotes hierarchy.
   */
  size?: keyof typeof headingSizes

  /**
   * The Dracula UI color to be applied to the Heading.
   */
  color?: keyof typeof headingColors

  /** The HTML element to be used */
  as?: keyof HTMLElementTagNameMap
}

/**
 * Heading is used to display headlines and other forms of hierarchical Text.
 *
 * Headings are similar to the base Text component, but restricted to certain
 * sizes and font weight.
 */
export const Heading: React.FC<HeadingProps> = (props: HeadingProps) => {
  const { size, color, as, ...htmlProps } = props

  const finalProps = {
    ...htmlProps,
    className: cx(
      `drac-heading`,
      props.className,
      headingSizes[size ?? 'xl'],
      headingColors[color ?? 'white'],
      ...paddingMixin(props),
      ...marginMixin(props)
    )
  }

  return React.createElement(as ?? 'h2', cleanProps(finalProps), props.children)
}

Heading.displayName = 'Heading'
