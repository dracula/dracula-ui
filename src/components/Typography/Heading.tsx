import cx from 'classnames'
import { mapValues } from 'lodash'
import React, { HTMLAttributes } from 'react'
import { colors } from '../../base/colors'
import { spacingClasses, SpacingPropType } from '../../base/spacing'

export const headingSizes = {
  'heading-1': 'drac-heading-1',
  'heading-2': 'drac-heading-2',
  'heading-3': 'drac-heading-3',
  'heading-4': 'drac-heading-4',
  'heading-5': 'drac-heading-5',
  'heading-6': 'drac-heading-6'
}

export const headingColors = mapValues(colors, (className) => {
  return className.replace('-bg-', '-text-')
})

/** Heading Props */
export interface HeadingProps extends HTMLAttributes<HTMLHeadingElement> {
  /**
   * The size and type of Heading to be used.
   * Denotes hierarchy.
   */
  size?: keyof typeof headingSizes

  /**
   * The Dracula UI color to be applied to the Heading.
   */
  color?: keyof typeof headingColors

  /**
   * Controls the spacing between the Heading component and its parent and siblings.
   */
  spacing?: SpacingPropType
}

/**
 * Heading is used to display headlines and other forms of hierarchical Text.
 *
 * Headings are similar to the base Text component, but restricted to certain
 * sizes and font weight.
 */
export const Heading: React.FC<HeadingProps> = (props: HeadingProps) => {
  const tag = {
    'heading-1': 'h1',
    'heading-2': 'h2',
    'heading-3': 'h3',
    'heading-4': 'h4',
    'heading-5': 'h5',
    'heading-6': 'h6'
  }

  const size = tag[props.size ?? 'heading-1']
  const finalProps = {
    ...props,
    className: cx(
      `drac-heading`,
      headingSizes[props.size ?? 'heading-1'],
      headingColors[props.color ?? 'white'],
      ...spacingClasses(props.spacing ?? 'none')
    )
  }

  return React.createElement(size, finalProps, props.children)
}

Heading.displayName = 'Heading'
