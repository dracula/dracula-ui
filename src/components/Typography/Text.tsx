import React, { HTMLAttributes } from 'react'
import { mapValues } from 'lodash'
import cx from 'classnames'
import { spacingClasses, SpacingPropType } from '../../base/spacing'
import { baseTextColors } from '../../base/colors'

export const textSizes = {
  large: 'drac-text-lg',
  medium: 'drac-text',
  small: 'drac-text-sm',
  xsmall: 'drac-text-xs'
}

export const textWeights = {
  normal: 'drac-text',
  semibold: 'drac-text-semibold',
  bold: 'drac-text-bold'
}

export const textColors = mapValues(baseTextColors, (className) => {
  return className.replace('-bg-', '-text-')
})

/** Text Props */
export interface TextProps extends HTMLAttributes<HTMLSpanElement> {
  /**
   * Controls the size of the text based on pre-configured Dracula UI sizes.
   * Options: `xsmall`, `small`, `medium`, `large`
   */
  size?: keyof typeof textSizes

  /**
   * Controsl the weight of the text.
   * Options: `normal`, `semibold`, `bold`.
   */
  weight?: keyof typeof textWeights

  /**
   * Controls the color of the text
   */
  color?: keyof typeof textColors

  /**
   * Controls the spacing between the Text component and its parent and siblings.
   */
  spacing?: SpacingPropType
}

/**
 * Text is the base component for any sort of text.
 *
 * Consumers of this component can control, the color, size,
 * weight, and spacing of Text.
 *
 * Use this component for generic, and non-hierarchical text that is
 * to be displayed on a page, or as part of other complex components
 * or UI patterns
 */
export const Text = (props: TextProps) => {
  const finalProps = {
    ...props,

    className: cx(
      `drac-text`,
      textSizes[props.size ?? 'medium'],
      textWeights[props.weight ?? 'normal'],
      textColors[props.color ?? 'white'],
      spacingClasses(props.spacing)
    )
  }

  return <span {...finalProps}>{props.children}</span>
}

/**
 * Paragraph is a semantic component used for blocks of text with
 * semantic meaning.
 *
 * Paragraph accepts all the same customization options as Text.
 */
export const Paragraph: React.FC<TextProps> = (props: TextProps) => {
  const finalProps = {
    ...props,

    className: cx(
      `drac-text`,
      textSizes[props.size ?? 'medium'],
      textWeights[props.weight ?? 'normal'],
      textColors[props.color ?? 'white'],
      spacingClasses(props.spacing ?? 'smallY')
    )
  }

  return <p {...finalProps}>{props.children}</p>
}

Text.displayName = 'Text'
