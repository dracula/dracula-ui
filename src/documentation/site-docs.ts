import {
  ComponentExample,
  renderSnapshot,
  SnapshotBuilder
} from '../story-helpers/render-component'
import React from 'react'
import fs from 'fs-extra'
import { pretty } from '../story-helpers/pretty'
import { getDocGen } from '../documentation/docgen/doc-generator'
import { toDSP } from '../documentation/dsp/component-generator'
import puppeteer from 'puppeteer'
import { componentScreenshot } from './screenshots/component-screenshot'

interface Documentation {
  basicUsage: () => SnapshotBuilder
  variations: () => Array<SnapshotBuilder>
}

jest.setTimeout(10000)

export function siteDocs<T>(
  componentClass: React.FunctionComponent<T>,
  documentation: Documentation
) {
  const name = componentClass.displayName ?? componentClass.name
  const dspPath = `${process.cwd()}/dsp`

  describe(`Site: ${name}`, () => {
    let browser: puppeteer.Browser

    beforeEach(async () => {
      browser = await puppeteer.launch({ headless: true, devtools: true })
    })

    afterEach(async () => {
      await browser.close()
    })

    let examples: Record<string, ComponentExample> = {}

    afterAll(async () => {
      const docGen = getDocGen(name)
      const dsp = toDSP(name, examples, docGen)

      await fs.mkdirp(`${dspPath}/data/components/`).catch(() => {
        // folder already exists
      })

      const path = `${dspPath}/data/components/${name}.json`
      await fs.writeFile(path, pretty(JSON.stringify(dsp), 'json'))
    })

    test('Basic Usage', async () => {
      const variation = documentation.basicUsage()
      const snapshot = renderSnapshot(
        variation.title,
        variation.comp,
        variation.docs
      )
      expect(snapshot).toMatchSnapshot()

      const [screenshot, svg] = await componentScreenshot(
        browser,
        snapshot,
        null,
        name
      )

      snapshot.screenshot = screenshot
      snapshot.svgData = svg

      examples = {
        ...examples,
        basic: snapshot
      }
    })

    const variations = documentation.variations()

    variations.forEach((variation) => {
      test(variation.title, async () => {
        const snapshot = renderSnapshot(
          variation.title,
          variation.comp,
          variation.docs
        )
        expect(snapshot).toMatchSnapshot()
        const [screenshot, svg] = await componentScreenshot(
          browser,
          snapshot,
          variation,
          name
        )

        snapshot.screenshot = screenshot
        snapshot.svgData = svg

        examples = {
          ...examples,
          [variation.title]: snapshot
        }
      })
    })
  })
}
