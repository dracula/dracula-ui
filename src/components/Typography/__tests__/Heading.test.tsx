import { Heading } from '@/components/Typography/Heading'
import {
  buildSnapshot,
  SnapshotBuilder
} from '@/story-helpers/render-component'
import { siteDocs } from '@/documentation/site-docs'
import {
  HeadingSizes,
  HeadingBasic,
  HeadingColors
} from '@/components/Typography/__stories__/Heading.stories'

siteDocs(Heading, {
  basicUsage() {
    return buildSnapshot('Usage', HeadingBasic)
  },
  variations() {
    return [
      buildSnapshot('Sizes', HeadingSizes),
      buildSnapshot('Colors', HeadingColors)
    ] as SnapshotBuilder[]
  }
})
