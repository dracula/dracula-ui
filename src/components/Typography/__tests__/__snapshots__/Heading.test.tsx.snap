// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Site: Heading Basic Usage 1`] = `
Object {
  "docs": "",
  "html": "<h1 class=\\"drac-heading drac-heading-1 drac-text-white drac-spacing-none\\">
  Heading
</h1>
",
  "react": "<Heading>Heading</Heading>;
",
  "title": "Usage",
}
`;

exports[`Site: Heading Colors 1`] = `
Object {
  "docs": "",
  "html": "<div>
  <div style=\\"margin-bottom: 20px\\">
    <h1
      color=\\"white\\"
      class=\\"drac-heading drac-heading-1 drac-text-white drac-spacing-none\\"
    >
      white
    </h1>
  </div>
  <div style=\\"margin-bottom: 20px\\">
    <h1
      color=\\"cyan\\"
      class=\\"drac-heading drac-heading-1 drac-text-cyan drac-spacing-none\\"
    >
      cyan
    </h1>
  </div>
  <div style=\\"margin-bottom: 20px\\">
    <h1
      color=\\"green\\"
      class=\\"drac-heading drac-heading-1 drac-text-green drac-spacing-none\\"
    >
      green
    </h1>
  </div>
  <div style=\\"margin-bottom: 20px\\">
    <h1
      color=\\"orange\\"
      class=\\"drac-heading drac-heading-1 drac-text-orange drac-spacing-none\\"
    >
      orange
    </h1>
  </div>
  <div style=\\"margin-bottom: 20px\\">
    <h1
      color=\\"pink\\"
      class=\\"drac-heading drac-heading-1 drac-text-pink drac-spacing-none\\"
    >
      pink
    </h1>
  </div>
  <div style=\\"margin-bottom: 20px\\">
    <h1
      color=\\"purple\\"
      class=\\"drac-heading drac-heading-1 drac-text-purple drac-spacing-none\\"
    >
      purple
    </h1>
  </div>
  <div style=\\"margin-bottom: 20px\\">
    <h1
      color=\\"red\\"
      class=\\"drac-heading drac-heading-1 drac-text-red drac-spacing-none\\"
    >
      red
    </h1>
  </div>
  <div style=\\"margin-bottom: 20px\\">
    <h1
      color=\\"yellow\\"
      class=\\"drac-heading drac-heading-1 drac-text-yellow drac-spacing-none\\"
    >
      yellow
    </h1>
  </div>
  <div style=\\"margin-bottom: 20px\\">
    <h1
      color=\\"purpleCyan\\"
      class=\\"drac-heading drac-heading-1 drac-text-purple-cyan drac-spacing-none\\"
    >
      purpleCyan
    </h1>
  </div>
  <div style=\\"margin-bottom: 20px\\">
    <h1
      color=\\"yellowPink\\"
      class=\\"drac-heading drac-heading-1 drac-text-yellow-pink drac-spacing-none\\"
    >
      yellowPink
    </h1>
  </div>
  <div style=\\"margin-bottom: 20px\\">
    <h1
      color=\\"cyanGreen\\"
      class=\\"drac-heading drac-heading-1 drac-text-cyan-green drac-spacing-none\\"
    >
      cyanGreen
    </h1>
  </div>
  <div style=\\"margin-bottom: 20px\\">
    <h1
      color=\\"pinkPurple\\"
      class=\\"drac-heading drac-heading-1 drac-text-pink-purple drac-spacing-none\\"
    >
      pinkPurple
    </h1>
  </div>
  <div style=\\"margin-bottom: 20px\\">
    <h1
      color=\\"animated\\"
      class=\\"drac-heading drac-heading-1 drac-text-animated drac-spacing-none\\"
    >
      animated
    </h1>
  </div>
</div>
",
  "react": "<div>
  <div style={{ marginBottom: 20 }}>
    <Heading color=\\"white\\">white</Heading>
  </div>
  <div style={{ marginBottom: 20 }}>
    <Heading color=\\"cyan\\">cyan</Heading>
  </div>
  <div style={{ marginBottom: 20 }}>
    <Heading color=\\"green\\">green</Heading>
  </div>
  <div style={{ marginBottom: 20 }}>
    <Heading color=\\"orange\\">orange</Heading>
  </div>
  <div style={{ marginBottom: 20 }}>
    <Heading color=\\"pink\\">pink</Heading>
  </div>
  <div style={{ marginBottom: 20 }}>
    <Heading color=\\"purple\\">purple</Heading>
  </div>
  <div style={{ marginBottom: 20 }}>
    <Heading color=\\"red\\">red</Heading>
  </div>
  <div style={{ marginBottom: 20 }}>
    <Heading color=\\"yellow\\">yellow</Heading>
  </div>
  <div style={{ marginBottom: 20 }}>
    <Heading color=\\"purpleCyan\\">purpleCyan</Heading>
  </div>
  <div style={{ marginBottom: 20 }}>
    <Heading color=\\"yellowPink\\">yellowPink</Heading>
  </div>
  <div style={{ marginBottom: 20 }}>
    <Heading color=\\"cyanGreen\\">cyanGreen</Heading>
  </div>
  <div style={{ marginBottom: 20 }}>
    <Heading color=\\"pinkPurple\\">pinkPurple</Heading>
  </div>
  <div style={{ marginBottom: 20 }}>
    <Heading color=\\"animated\\">animated</Heading>
  </div>
</div>;
",
  "title": "Colors",
}
`;

exports[`Site: Heading Sizes 1`] = `
Object {
  "docs": "",
  "html": "<div>
  <div style=\\"margin-bottom: 20px\\">
    <h1
      color=\\"white\\"
      class=\\"drac-heading drac-heading-1 drac-text-white drac-spacing-none\\"
    >
      heading-1
    </h1>
  </div>
  <div style=\\"margin-bottom: 20px\\">
    <h2
      color=\\"white\\"
      class=\\"drac-heading drac-heading-2 drac-text-white drac-spacing-none\\"
    >
      heading-2
    </h2>
  </div>
  <div style=\\"margin-bottom: 20px\\">
    <h3
      color=\\"white\\"
      class=\\"drac-heading drac-heading-3 drac-text-white drac-spacing-none\\"
    >
      heading-3
    </h3>
  </div>
  <div style=\\"margin-bottom: 20px\\">
    <h4
      color=\\"white\\"
      class=\\"drac-heading drac-heading-4 drac-text-white drac-spacing-none\\"
    >
      heading-4
    </h4>
  </div>
  <div style=\\"margin-bottom: 20px\\">
    <h5
      color=\\"white\\"
      class=\\"drac-heading drac-heading-5 drac-text-white drac-spacing-none\\"
    >
      heading-5
    </h5>
  </div>
  <div style=\\"margin-bottom: 20px\\">
    <h6
      color=\\"white\\"
      class=\\"drac-heading drac-heading-6 drac-text-white drac-spacing-none\\"
    >
      heading-6
    </h6>
  </div>
</div>
",
  "react": "<div>
  <div style={{ marginBottom: 20 }}>
    <Heading color=\\"white\\" size=\\"heading-1\\">
      heading-1
    </Heading>
  </div>
  <div style={{ marginBottom: 20 }}>
    <Heading color=\\"white\\" size=\\"heading-2\\">
      heading-2
    </Heading>
  </div>
  <div style={{ marginBottom: 20 }}>
    <Heading color=\\"white\\" size=\\"heading-3\\">
      heading-3
    </Heading>
  </div>
  <div style={{ marginBottom: 20 }}>
    <Heading color=\\"white\\" size=\\"heading-4\\">
      heading-4
    </Heading>
  </div>
  <div style={{ marginBottom: 20 }}>
    <Heading color=\\"white\\" size=\\"heading-5\\">
      heading-5
    </Heading>
  </div>
  <div style={{ marginBottom: 20 }}>
    <Heading color=\\"white\\" size=\\"heading-6\\">
      heading-6
    </Heading>
  </div>
</div>;
",
  "title": "Sizes",
}
`;
